core/src/main/java/org/apache/iceberg/io/CachingFileIO.java

```diff
@@ -0,0 +1,404 @@
+
+      Caffeine<Object, Object> builder = Caffeine.newBuilder();
+      if (expireAfterAccessMs > 0) {
+        builder = builder.expireAfterAccess(Duration.ofMillis(expireAfterAccessMs));
+      }
+
+      this.cache = builder.maximumWeight(maxTotalBytes)
+          .weigher((Weigher<String, byte[]>) (key, value) -> value.length)
+          .recordStats()
+          .build();
+    }
+
+    public long expireAfterAccess() {
+      return expireAfterAccessMs;
+    }
+
+    public long maxContentLength() {
+      return maxContentLength;
+    }
+
+    public long maxTotalBytes() {
+      return maxTotalBytes;
+    }
+
+    public Cache<String, byte[]> cache() {
+      return cache;
+    }
+
+    public CacheStats stats() {
+      return cache.stats();
+    }
+
+    @Override
+    public String toString() {
+      return getClass().getSimpleName() + '{' +
+          "expireAfterAccessMs=" + expireAfterAccessMs + ", " +
+          "maxContentLength=" + maxContentLength + ", " +
+          "maxTotalBytes=" + maxTotalBytes + ", " +
+          "cacheStats=" + cache.stats() + '}';
+    }
+  }
+
+  private static class CachingInputFile implements InputFile {
+    private final ContentCache contentCache;
+    private final InputFile wrappedInputFile;
+
+    private CachingInputFile(ContentCache cache, InputFile inFile) {
+      this.contentCache = cache;
+      this.wrappedInputFile = inFile;
+    }
+
+    @Override
+    public long getLength() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return (buf != null) ? buf.length : wrappedInputFile.getLength();
+    }
+
+    @Override
+    public SeekableInputStream newStream() {
+      try {
+        if (getLengthChecked() <= contentCache.maxContentLength()) {
+          return cachedStream();
+        }
+
+        return wrappedInputFile.newStream();
+      } catch (FileNotFoundException e) {
+        throw new NotFoundException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      } catch (IOException e) {
+        throw new RuntimeIOException(e, "Failed to open input stream for file: %s", wrappedInputFile.location());
+      }
+    }
+
+    @Override
+    public String location() {
+      return wrappedInputFile.location();
+    }
+
+    @Override
+    public boolean exists() {
+      byte[] buf = contentCache.cache.getIfPresent(location());
+      return buf != null || wrappedInputFile.exists();
+    }
+
+    public long getLengthChecked() throws IOException {
+      try {
+        return getLength();
+      } catch (RuntimeIOException ex) {
+        throw ex.getCause();
+      }
+    }
+
+    private byte[] getFileContent() {
+      try {
+        int totalByteRead = 0;
+        boolean eos = false;
+        int fileLength = (int) getLength();
+        byte[] buf = new byte[fileLength];
+        SeekableInputStream stream = wrappedInputFile.newStream();
+        while (totalByteRead < fileLength && !eos) {
```

Comment:
Can this use `ByteStreams.readFully` instead of implementing its own `readFully` method?

Reply:
Refactored to use `IOUtil.readRemaining`